lib/advisory_db.rb

```diff
@@ -0,0 +1,101 @@
+class AdvisoryDb
+  def self.update?
+    new_sha = `git ls-remote http://github.com/rubysec/ruby-advisory-db refs/heads/master`.split.first
+    if !Dir.exist?('tmp/advisories')
+      Dir.mkdir('tmp/advisories')
+      Dir.chdir(Rails.root.join('tmp', 'advisories'))
+      if system "git clone  --depth 1 http://github.com/rubysec/ruby-advisory-db #{Dir.pwd}/ruby-advisory-db"
+        File.write('advisory_db_head_sha', new_sha) # TODO: store new sha after updating db
+      end
+    else
+      Dir.chdir(Rails.root.join('tmp', 'advisories'))
+      clone_sha = File.open('advisory_db_head_sha', &:readline)
+      return false unless new_sha != clone_sha
+      advisory_db_path = Rails.root.join('tmp', 'advisories', 'ruby-advisory-db')
+      url = "http://github.com/rubysec/ruby-advisory-db"
+      File.write('advisory_db_head_sha', new_sha) if system "rm -rf #{advisory_db_path} && git clone  --depth 1 #{url} #{advisory_db_path}"
+    end
+  end
+
+  def self.find_vulnerability(version, ignored_cves)
+    version_advisory_files = []
+    Dir.chdir(Rails.root.join('tmp', 'advisories', 'ruby-advisory-db'))
+    if Dir.exist?("gems/#{version.rubygem.name}")
+      Dir.glob("gems/#{version.rubygem.name}/*.yml") do |advisory|
+        version_advisory_files << advisory if contain?(advisory, version)
+      end
+    end
+    find_details(version, version_advisory_files, ignored_cves)
+  end
+
+  def self.find_details(_version, version_advisory_files, ignored_cves)
+    details = []
+    version_advisory_files.each do |advisory|
+      File.open(advisory) do |_f|
+        advisory = YAML.load_file(advisory)
+        details << { title: advisory['title'], url: advisory['url'] } unless ignored_cves.include? advisory['cve']
+      end
+    end
+    details
+  end
+
+  def self.contain?(advisory, version)
+    advisory = YAML.load_file(advisory)
+    version_array = Array.new(1, version.number)
+    remove_unaffected_versions(version_array, (advisory['patched_versions'].to_a + advisory['unaffected_versions'].to_a).flatten).present?
```

Comment:
Do you need `flatten` since `+` will merge the 2 arrays and return a new array? If each left/right side array is multi-dimension, then `flatten` is needed.